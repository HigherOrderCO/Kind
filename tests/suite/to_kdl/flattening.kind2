List <a> : Type
  List.nil <a> : List
  List.cons <a> (head: a) (tail: List a) : List

Ora.Ora.ora <a> (h: List a) : List a
Ora.Ora.ora List.nil         = List.nil
Ora.Ora.ora (List.cons x xs) = List.cons x xs

Null <a> (h: List (List a)) : List a
Null (List.cons (List.cons y ys) (List.cons x (List.cons z zs))) = List.nil
Null (List.cons (List.nil) (List.cons (List.cons x ys) xs))      = List.nil