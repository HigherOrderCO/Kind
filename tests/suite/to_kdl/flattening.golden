// List.nil #List.nil -(a: Type) : (List _)
ctr {List.nil}

// List.cons #List.cons -(a: Type) (head: a) (tail: (List a)) : (List _)
ctr {List.cons head tail}

// Ora.Ora.ora #Ora.Ora.ora -(a: Type) (h: (List a)) : (List a)
fun (Ora.Ora.ora h) {
  (Ora.Ora.ora {List.nil}) = {List.nil}
  (Ora.Ora.ora {List.cons x xs}) = {List.cons x xs}
}

fun (Null0_0_ x0 x1 x2 x3) {
  (Null0_0_ ~ ~ ~ {List.cons ~ ~}) = {List.nil}
}

fun (Null0_1_ x0 x1) {
  (Null0_1_ {List.cons ~ ~} ~) = {List.nil}
}

fun (Null0_ x0 x1) {
  (Null0_ {List.cons y ys} {List.cons x .0}) = (Null0_0_ y ys x .0)
  (Null0_ {List.nil} {List.cons .0 xs}) = (Null0_1_ .0 xs)
}

// Null #Null -(a: Type) (h: (List (List a))) : (List a)
fun (Null h) {
  (Null {List.cons .0 .1}) = (Null0_ .0 .1)
}

